tests/actions/scan_test.py

```diff
@@ -106,12 +106,15 @@ def test_does_not_write_state_when_dry_run(self, mock_file_operations):
 
         assert not mock_file_operations.write.called
 
-    def test_does_not_write_state_when_scan_head(self, mock_file_operations):
+    def test_scan_head_and_does_not_write_state_when_scan_head(self, mock_file_operations):
         with self.setup_env(
             SecretsCollection(),
             '--scan-head',
         ) as args:
-            assert scan_repo(args) == 0
+            with mock.patch('detect_secrets_server.actions.scan.tracked_repo_factory') as mock_repo_factory:
+                assert scan_repo(args) == 0
+                mock_repo = mock_repo_factory.return_value.load_from_file.return_value
+                mock_repo.scan.assert_called_once_with(exclude_files_regex=None, exclude_lines_regex=None, scan_head=True)
```

Comment:
A better test case would be to leverage `setup_env` to mock scan results, then check the logger to see whether it complained about finding secrets (while at the same time, making sure that no state is written). See `test_alerts_on_secrets_found` for more details.

With regards to style, you can avoid hadouken code by avoiding nesting `with` statements. See `setup_env` for an example.